### PR TITLE
Frontend refactor, and some documentation

### PR DESCRIPTION
**Pull request by ezyang**
Frontend refactor, and some documentation.
    
    - BC BREAKING: export now also takes a mandatory file-ish argument, specifying
      the file to export the protobuf to.  I rewrote the tests to use BytesIO to
      get out the string so they could parse it again.
    
    - BC BREAKING: export no longer returns the tensors that were computed.  To
      get these, use the internal _export function.
    
    - Multiple inputs to models are now supported by passing a tuple to input.
      (Old API of a single Variable still works.)
    
    - Keyword arguments to models are now supported via kwargs keyword arg.
    
    - Renamed embed_params to export_params, and it now defaults to True.
    
    - Toffee tests now live in their own test_toffee.py file.  I had to
      rename a pile of expect files for this.
    
    - Removed defunct torch.toffee imports from autograd to solve module import
      cycle.
    
    - Helper function _with_file_like to abstract over opening file-ish arguments,
      taken from torch.save()

### CODE REVIEW

File: torch/toffee.py [hunk 1]
```diff
@@ -1,24 +1,53 @@
...
+    once in order to get a trace of its execution to be exported; at the
+    moment, it does not support dynamic models (e.g., RNNs.)
 
+    See also: :ref:`toffee-export`
 
-def export(model, input, embed_params):
+    Arguments:
+        model (torch.nn.Module): the model to be exported.
+        args (torch.autograd.Variable or tuple of variables): the inputs to
+            the model, e.g., such that ``model(*args)`` is a valid invocation
```
Comment (apaszke) [inline torch/toffee.py]: it doesn't have to be valid without `kwargs`

File: torch/toffee.py [hunk 2]
```diff
@@ -1,24 +1,53 @@
...
+    if export_params:
         proto = trace.export(model.state_dict().values())
-    # TODO: a way to print the proto
-    return proto, torch_out
+    else:
+        proto = trace.export()
+    torch.serialization._with_file_like(f, "wb", lambda f: f.write(proto))
+    return torch_out
+    # NB: It's very important that trace dies at the end of this function;
+    # otherwise you can't retrace the model.
```
Comment (apaszke) [inline torch/toffee.py:55]: Don't add this comment. It's going to change soon

Comment (ezyang) [reply torch/toffee.py:55]: Probably not for the release though.

Comment (apaszke) [reply torch/toffee.py:55]: So what? It's better to not scatter misleading comments over the code

**Pull request merged**